### ISSUE TITLE
Support for 18.04 Server

### ISSUE THREAD

**Issue by sam0737** [saveriomiroddi/zfs-installer#11]
Thank you for the great work and it works great for desktop version.

Unfortunately it doesn't work with server edition (ubuntu-18.04.3-live-server-amd64.iso)
/lib/modules is mounted as readonly to the installation media and installation of the zfs kmod failed.

**Comment by saveriomiroddi** [comment]
Hello! Good point. I'll check and figure out if I can do something :smile:.

**Comment by saveriomiroddi** [comment]
What's the procedure you use to install Ubuntu server (via terminal, I assume)?

I didn't find any option in the Ubuntu Server installater to drop into a commandline (which is required to start the script), so I wonder how you got there.

On top of my head, to solve the problem, I can think of:

1. disabling the 0.8 module installation. this will make encryption and TRIM unavailable. the first may not be a big deal on a server context; not sure about the second
2. generally provide a script to install ubuntu server via debootstrap

unless of course, there is some exotic way to install modules in alternative directories, or to remount `/lib/modules` in r/w mode.

**Comment by kongkrit** [comment]
I’m nowhere near Linux guru, but would an Ubuntu minimal iso help? It allows you to pick and choose package, and I think maybe able to drop you to terminal.

Please see https://help.ubuntu.com/community/Installation/MinimalCD

**Comment by saveriomiroddi** [comment]
> I’m nowhere near Linux guru, but would an Ubuntu minimal iso help? It allows you to pick and choose package, and I think maybe able to drop you to terminal.
> 
> Please see https://help.ubuntu.com/community/Installation/MinimalCD

Thanks for the participation :smile: 

So, I've never tried it, if somebody can post the details (I didn't get any feedback for my request), I can work on it.

In general terms, pretty much any Debian-based (and in turn, any Ubuntu-based) installation is supported by zfs-installer, via:

```
- ZFS_OS_INSTALLATION_SCRIPT : path of a script to execute instead of Ubiquity (see dedicated section below)
```

Essentially, the user creates a script that performs the installation via `debootstrap`, and zfs-installer does all the required gruntwork.

I'm closing the issue. I'll reopen it if somebody's interested and reports a specific issue or requires a specific update of the code.

**Comment by kongkrit** [comment]
Hi saveriomiroddi,

Thanks for your kind reply. I was trying to hack ubuntu 18.04.3 live server ISO to work with your script, but given my low level of expertise, got stuck.

I was trying in virtualbox with 2 hard disks of equal size, to be formatted as ZFS mirror.
I got as far as getting your script to prepare/parition the disks successfully, and got it to launch _subiquity_ (the server's text-based ubiquity), but unfortunately, during the step:

select /dev/zd16p1 and format it as ext4 and mount to /

there is no such option for subiquity. and i'm stuck. Here's the log of what I did so far, in case you're interested:

1. boot ubuntu server 18.04.3 live cd and choose "boot and install with HWE kernel" then "Install Ubuntu Server" at welcome screen, type Alt+F2
2. passwd to change ubuntu-server password -- the current password is empty - so i can ssh in
3. sudo apt install openssh-server -y
4. connect via ssh ubuntu-server@ipaddress and then sudo -i
5. systemctl stop systemd-udevd*  -- to stop it grabbing /lib/modules
6. create rw ramdisk in place of readonly /lib/modules:
mkdir /lib/modules-ramdisk && \
mount -t tmpfs -o rw,size=1G tmpfs /lib/modules-ramdisk && \
cp -r /lib/modules /lib/modules-ramdisk && \
umount /lib/modules && \
mount -t tmpfs -o rw,size=1G tmpfs /lib/modules && \
cp -r /lib/modules-ramdisk /lib/modules && \
umount /lib/modules-ramdisk && \
rmdir /lib/modules-ramdisk
7. apt install linux-generic-hwe-18.04 -y  -- gets kernel v 5 ready
8. nano /etc/apt/sources.list -- add sources 
deb cdrom:[Ubuntu-Server 18.04.3 LTS _Bionic Beaver_ - Release amd64 (20190805)]/ bionic main restricted
deb http://archive.ubuntu.com/ubuntu/ bionic main restricted universe multiverse
deb http://archive.ubuntu.com/ubuntu/ bionic-updates main restricted universe multiverse
deb http://archive.ubuntu.com/ubuntu/ bionic-backports main restricted universe multiverse
deb http://security.ubuntu.com/ubuntu/ bionic-security main restricted universe multiverse
9. apt update
10. ln -s /snap/bin/subiquity /usr/bin/ubiquity <-- make your script launch subiquity
11. wget -qO- https://git.io/JelI5 | bash /dev/stdin

And that's where i'm stuck at. So close. :-(

**Comment by saveriomiroddi** [comment]
> Hi saveriomiroddi,
> 
> Thanks for your kind reply. I was trying to hack ubuntu 18.04.3 live server ISO to work with your script, but given my low level of expertise, got stuck.

Ah! I see. I'm going to reopen the issue, and work on it over the next week :smile:

**Comment by saveriomiroddi** [comment]
> Hi saveriomiroddi,
> 
> Thanks for your kind reply. I was trying to hack ubuntu 18.04.3 live server ISO to work with your script, but given my low level of expertise, got stuck.

So! At a quick glance, a couple of things:

1. you don't need to create two tmpfs filesystems; `/tmp` is already one :smile:. I need to test it, but it's possible that you can just copy the `/lib/modules` somewhere under `/tmp`, unmount the former, create a symlink, and voilà :smile:
2. the problem is very likely caused by the missing udev subsystem. Udev is crucial for managing devices, so what's happening, I reckon, is that when the volume is created, the corresponding block device (under `/dev`) is not created as well. If this is the case, just restarting the daemons will do the trick.

I'm going to add the support for this over the next days.

**Comment by saveriomiroddi** [comment]
> Hi saveriomiroddi,
> 
> Thanks for your kind reply. I was trying to hack ubuntu 18.04.3 live server ISO to work with your script, but given my low level of expertise, got stuck.

At a *very* quick test, the strategy works:

```sh
# output not printed here
$ systemctl stop systemd-udevd*
$ cp -Rv /lib/modules /tmp
$ umount /lib/modules
$ rm -r /lib/modules
$ ln -s /tmp/modules /lib
$ systemctl start systemd-udevd
$ systmctl list-units systemd-udevd
# units active!
```

I didn't try to follow up, though. There are still 700+ MiB free on `/tmp`, and I think the `/lib/modules` files are normal files, so in theory it should work.

**Comment by saveriomiroddi** [comment]
After a lot of pains, it turns out that the problem is with Subiquity, which doesn't allow installing in virtual devices (which is what the script uses):

- [Cannot make Subiquity see zfs zvol block device, like I could with Ubiquity](https://askubuntu.com/questions/1060041/cannot-make-subiquity-see-zfs-zvol-block-device-like-i-could-with-ubiquity)
- [Cannot install subiquity on virtual devices, including zvols](https://bugs.launchpad.net/subiquity/+bug/1811037)

This can be worked around by using a temporary disk (a third one, in your case), and `rsync`ing to the root rpool.

I will add support, but I need to check carefully how.

**Comment by saveriomiroddi** [comment]
An alternative is to reserve space for an ext4 partition, install there, sync, remove it, and optionally expand the zfs pool partition, and expand the zfs pool. It's quite some work :-)

**Comment by kongkrit** [comment]
> After a lot of pains, it turns out that the problem is with Subiquity, which doesn't allow installing in virtual devices (which is what the script uses):
> 
> * [Cannot make Subiquity see zfs zvol block device, like I could with Ubiquity](https://askubuntu.com/questions/1060041/cannot-make-subiquity-see-zfs-zvol-block-device-like-i-could-with-ubiquity)
> * [Cannot install subiquity on virtual devices, including zvols](https://bugs.launchpad.net/subiquity/+bug/1811037)
> 
> This can be worked around by using a temporary disk (a third one, in your case), and `rsync`ing to the root rpool.

Thanks again for your kind help.
So, according to above, I'd temporarily need 3 hard disks, let your script create partitions on the first 2, install ubuntu to the 3rd, and then rsync all the files from the 3rd HD to the ZFS filesystem created by your script on the first 2 drives?

Doesn't sound too bad, actually, but if I understand correctly, there are 3-4 partitions on the ZFS formatted disk, correct?
1. EFI
2. boot (bpool)
3. root (rpool)
(optional 4.) swap

boot and root can be rsync, but EFI is a straight copy, since it is not a ZFS volume. But how do we make sure the EFI and grub see the correct partitions after just rsync'ing the files over? (My lack of knowledge of linux booting process here).

Frankly, this sounds a lot simpler than the alternative in the next reply (messing with the partitions).

I'm learning a lot more about sysadmin stuff. Thanks a lot again!

**Comment by saveriomiroddi** [comment]
@sam0737 @kongkrit Support added. Have fun :smile: If you find any issue, let me know.

**Comment by kongkrit** [comment]
> @sam0737 @kongkrit Support added. Have fun 😄 If you find any issue, let me know.

Thanks a lot! I will give it a try ASAP and give you some feedback.
